[Quant] Add quantization 2.0 document (#2354)
pytorch/tutorials: prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst

This API covers most use cases relatively well, but the main problem is that this API is not fully extensible
with two main limitations:

-  Limitation around expressing quantization intentions for complicated operator patterns such as in the discussion of
   `issue-96288 <https://github.com/pytorch/pytorch/issues/96288>`__ to support ``conv add`` fusion with oneDNN library.
   It also requires some changes to current already complicated pattern matching code such as in the
   `PR-97122 <https://github.com/pytorch/pytorch/pull/97122>`__ to support ``conv add`` fusion.
-  Limitation around supporting user's advanced quantization intention to quantize their model. For example, if backend
   developer only want to quantize inputs and outputs when the ``linear`` has a third input, it requires co-work from quantization
   team and backend developer.

To address these scalability issues, 
`Quantizer <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L160>`__
is introduced for quantization in PyTorch 2.0 export. ``Quantizer`` is a class that users can use to

> programmably set the observer or fake quant objects for each node in the model graph. It adds flexibility

Reviewer: `programmably set the observer or fake quant objects for each node in the model graph.` I think this is out of date, we should say `programmably set the quantization specifications for input and output of each node in the model graph.`

Author: Thanks and Fixed.

to the quantization API and allows modeling users and backend developers to configure quantization programmatically.
This will allow users to express how they want an operator pattern to be observed in a more explicit
way by annotating the appropriate nodes. A backend specific quantizer inherited from base quantizer,

> some APIs need to be overrided:

Reviewer: ```suggestion
some methods that need to be implemented:
```

Author: Thanks and changed.

-  `annotate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L269>`__

>    is used to annotate nodes in the graph with observer or fake quant constructors to convey the desired way of quantization.

Reviewer: ```suggestion
   is used to annotate nodes in the graph with QuantizationAnnotation objects to convey the desired way of quantization.
```

Author: Thanks and changed.

> -  `validate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L721>`__

Reviewer: I think maybe we can skip this for now? I think we haven't figure out the detail for this one yet?

cc @kimishpatel what do you think?

Author: Thanks and temporarily removed the ``validate method``.

of how this intent is conveyed in the quantization workflow with annotation API.

-  Step 1: Identify the original floating point pattern in the FX graph. There are
   several ways to identify this pattern: User may use a pattern matcher (e.g. SubgraphMatcher)
   to match the operator pattern; User may go through the nodes from start to the end and compare
   the node's target type to match the operator pattern. In this example, we can use the
   `get_source_partitions <https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/fx/passes/utils/source_matcher_utils.py#L51>`__
   to match this pattern. The original floating point ``add`` pattern only contain a single ``add`` node. 

::

    add_partitions = get_source_partitions(gm.graph, [operator.add, torch.add])
    add_partitions = list(itertools.chain(*add_partitions.values()))
    for add_partition in add_partitions:
        add_node = add_partition.output_nodes[0]

-  Step 2: Define the ``QuantizationSpec`` for inputs and output of the pattern. ``QuantizationSpec``
   defines the ``data type``, ``qscheme``, and other quantization parameters about users' intent of

>    how to observer/quantize a tensor.

Reviewer: ```suggestion
   how to observe or fake quantize a tensor.
```

Author: Thanks and changed.

It is natural that users want to annotate a quantized model where quantization
parameters can be shared among some tensors explicitly. Two typical use cases are:

-  Example 1: One example is for ``add`` where having both inputs sharing quantization
   parameters makes operator implementation much easier. Without using of
   `SharedQuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L90>`__,
   we have to annotate ``add`` as example in above section 1, in which two inputs of ``add``
   has different quantization parameters.
-  Example 2: Another example is that of sharing quantization parameters between inputs and output.
   This typically results from operators such as ``maxpool``, ``average_pool``, ``concat`` etc.

``SharedQuantizationSpec`` is designed for this use case to annotate tensors whose quantization
parameters are shared with other tensors. Input of ``SharedQuantizationSpec`` is an ``EdgeOrNode`` object which 
can be an input edge or an output value. 

-  Input edge is the connection between input node and the node consuming the input,

>    so it's a Tuple[Node, Node].

Reviewer: ```suggestion
   so it's a ``Tuple[Node, Node]``.
```

Author: Thanks and changed.

> -  Output value is an fx Node.

Reviewer: ```suggestion
-  Output value is an fx ``Node``.
```

Author: Thanks and changed.

Now, if we want to rewrite ``add`` annotation example with ``SharedQuantizationSpec`` to indicate
two input tensors as sharing quantization parameters. We can define its ``QuantizationAnnotation``
as this:

-  Step 1: Annotate input_act0 of ``add`` with ``QuantizationSpec``.
-  Step 2: Create a ``SharedQuantizationSpec`` object with input edge defined as ``(input_act0, add_node)`` which means to
   share the observer used for this edge. Then, user can annotate input_act1 with this ``SharedQuantizationSpec``

>    object.

Reviewer: nit: step 1 should be find the pattern, the same as the other examples, could you add that? same for section 3 and section 4

Author: Thanks. Add the ``pattern match`` as step1 for both section 2, 3, 4.

        _annotated=True,
    )

3. Annotate fixed qparams operators
--------------------------------------------------------

Another typical use case to annotate a quantized model is for tensors whose
quantization parmaters are known beforehand. For example, operator like ``sigmoid``, which has
predefined and fixed scale/zero_point at input and output tensors.
`FixedQParamsQuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L90>`__
is designed for this use case. To use ``FixedQParamsQuantizationSpec``, users need to pass in parameters
of ``scale`` and ``zero_point`` explicitly.

-  Step 1: Create ``FixedQParamsQuantizationSpec`` object with inputs of fixed ``scale``, ``zero_point`` value.
   These values will be used to create the ``quantize`` node and ``dequantize`` node in the convert phase.
-  Step 2: Annotate inputs and output to use this ``FixedQParamsQuantizationSpec`` object.

::

    act_qspec = FixedQParamsQuantizationSpec(
        dtype=torch.uint8,
        quant_min=0,
        quant_max=255,
        qscheme=torch.per_tensor_affine,

>         scale=2.0 / 256.0,
>         zero_point=128,

Reviewer: ```suggestion
        scale=1.0 / 256.0,
        zero_point=0,
```

I remember these have been updated

Author: Oh, yean. Thanks and updated.

        input_qspec_map={input_act: act_qspec},
        output_qspec=act_qspec,
        _annotated=True,
    )

4. Annotate tensor with derived quantization parameters
---------------------------------------------------------------

Another use case is to define the constraint for tensors whose quantization parameters are derived from other tensors.
For example, if we want to annotate a convolution node, and define the ``scale`` of its bias input tensor
as product of the activation tensor's ``scale`` and weight tensor's ``scale``. We can use
`DerivedQuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L102>`__
to annotate this conv node.

-  Step 1: Define ``derive_qparams_fn`` function, it accepts list of ``ObserverOrFakeQuantize`` (
   `ObserverBase <https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/ao/quantization/observer.py#L124>`__
   or `FakeQuantizeBase <https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/ao/quantization/fake_quantize.py#L60>`__)

>    as input. From each ``ObserverOrFakeQuantize`` object, user can get the ``scale``, ``zero point`` value. Combine these values together,
>    user can define its heuristic about how to derive new ``scale``, ``zero point`` value.

Reviewer: ```suggestion
   as input. From each ``ObserverOrFakeQuantize`` object, user can get the ``scale``, ``zero point`` value. User can define its heuristic about how to derive new ``scale``, ``zero point`` value based on the quantization parameters calculated from the observer or fake quant instances.
```

Author: Thanks and update the description.

        weight_obs_or_fq = obs_or_fqs[1]
        act_scale, act_zp = act_obs_or_fq.calculate_qparams()
        weight_scale, weight_zp = weight_obs_or_fq.calculate_qparams()
        return torch.tensor([act_scale * weight_scale]).to(torch.float32), torch.tensor([0]).to(torch.int32)

    bias_qspec = DerivedQuantizationSpec(
        derived_from=[(input_act, node), (weight, node)],
        derive_qparams_fn=derive_qparams_fn,
        dtype=torch.int32,
        quant_min=-2**31,
        quant_max=2**31 - 1,
        qscheme=torch.per_tensor_symmetric,
    )
    input_qspec_map = {input_act: act_quantization_spec, weight: weight_quantization_spec, bias: bias_qspec}
    node.meta["quantization_annotation"] = QuantizationAnnotation(
        input_qspec_map=input_qspec_map,
        output_qspec=act_quantization_spec,
        _annotated=True,
    )

5. A Toy Example with Resnet18 
--------------------------------------------------------

After above annotation methods defined with ``QuantizationAnnotation API``, we can now put them together to construct a ``BackendQuantizer``
and run a `toy example <https://gist.github.com/leslie-fang-intel/b78ed682aa9b54d2608285c5a4897cfc>`__

> with ``Torchvision Resnet18``. To better understand the final example, here are some basic concepts which are used:

Reviewer: ```suggestion
with ``Torchvision Resnet18``. To better understand the final example, here are the classes and utility functions that are used in the example:
```

Author: Thanks and changed.

- `QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
  consists of ``QuantizationSpec`` for activation, weight, and bias separately.

> - When annotating the model, methods of

Reviewer: ```suggestion
- When annotating the model,
```

Author: Thanks and changed.

>   `get_act_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#L9>`__,

Reviewer: nit: this is now changed to https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/utils.py#L10 and https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/utils.py#L23

Author: Thanks and the link are updated.

> (prototype) Quantization in PyTorch 2.0 Export Tutorial

Reviewer 2: Can we slap work in progress for this?

Author: Thanks. Add  `Work in Progress` in the title.

transformations on top of the ATen dialect graph. This approach is expected to
have significantly higher model coverage, better programmability, and
a simplified UX.

Prerequisites:
-----------------------

-  `Understanding of torchdynamo concepts in PyTorch <https://pytorch.org/docs/stable/dynamo/index.html>`__
-  `Understanding of the quantization concepts in PyTorch <https://pytorch.org/docs/master/quantization.html#quantization-api-summary>`__
-  `Understanding of FX Graph Mode post training static quantization <https://pytorch.org/tutorials/prototype/fx_graph_mode_ptq_static.html>`__
-  `Understanding of BackendConfig in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html?highlight=backend>`__
-  `Understanding of QConfigMapping in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html#set-up-qconfigmapping-that-satisfies-the-backend-constraints>`__

Previously in ``FX Graph Mode Quantization`` we were using ``QConfigMapping`` for users to specify how the model to be quantized
and ``BackendConfig`` to specify the supported ways of quantization in their backend.

> This API covers most use cases relatively well, but the main problem is that this API is not fully extensible

Reviewer 2: @jerryzh168 changes to new API is not so much about whether it covers the cases well or not, it is just a new way of thinking of the quantizaiton workflow in the context of PT2 export. So I dont think it is quite accurate to say that we built new workflow to cover tiny percentage of cases that werent working well.

THis is also the reason why i dont want to jump into documenting something that is not quite true to the reason of why new API exist. cc: @supriyar

Reviewer: I feel maybe we should mention (3) and (4) as well in https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c which are not limitations but more talking about motivations for the new API to have the complete picture?

Author: Thanks for the suggestions. section 3, 4 are added.

Reviewer 3: I don't think we should be documenting the details on the motivation for quantizer and annotation API to this level in this tutorial. Summarizing the motivation in 1-2 sentences should be good and the details can be left to the appendix or some other detailed doumentation/notes.

with two main limitations:

-  Limitation around expressing quantization intentions for complicated operator patterns such as in the discussion of
   `issue-96288 <https://github.com/pytorch/pytorch/issues/96288>`__ to support ``conv add`` fusion with oneDNN library.
   It also requires some changes to current already complicated pattern matching code such as in the

>    `PR-97122 <https://github.com/pytorch/pytorch/pull/97122>`__ to support ``conv add`` fusion.

Reviewer 2: This is an outcome of the new API, not necessarily the intention with which it was developed. So to say that these are two limitations for which we are developing new API is not quite accurate.

Author: Thanks. I think it's similar as previous comment. So as suggested by Jerry in here https://github.com/pytorch/tutorials/pull/2354#discussion_r1223618560. 2 more bullets are added to explain the motivations.

            super().__init__()
            self.linear = torch.nn.Linear(5, 10)

        def forward(self, x):
            return self.linear(x)

    example_inputs = (torch.randn(1, 5),)
    model = M().eval()

    # Step 1: Trace the model into an FX graph of flattened ATen operators
    exported_graph_module, guards = torchdynamo.export(
        model,
        *copy.deepcopy(example_inputs),
        aten_graph=True,
    )

    # Step 2: Insert observers or fake quantize modules
    quantizer = qq.QNNPackQuantizer()
    operator_config = qq.get_symmetric_quantization_config(is_per_channel=True)
    quantizer.set_global(operator_config)
    prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)

    # Step 3: Quantize the model
    convered_graph_module = convert_pt2e(prepared_graph_module)

    # Step 4: Lower Reference Quantized Model into the backend

Inside the Quantizer, we will use the ``QuantizationAnnotation API``
to convey user's intent for what quantization spec to use and how to
observe certain tensor values in the prepare step. Now, we will have a step-by-step
tutorial for how to use the ``QuantizationAnnotation API`` with different types of

> ``QuantizationSpec``.

Reviewer 2: Befor we introduce any of this, we need to explain what annotation API does. For that we first also have to explain quantization spec.

So the flow I imagine would be
```
Quantizer uses annotation API to convey quantization intent for different operators/patterns. Annotation API uses QuantizationSpec to convey intent of how a tensor will be quantized, e.g. dtype, bitwidth, min, max values, symmetric vs. asymmetric etc. See the spec definition at.... Furthermore it also allows quantizer to specify how a tensor value should be observed, e.g. minmax vs. histogram vs. something custom.

Quantization spec is then used to annotate nodes' output tensor or input tensors. Annotating input tensors is equivalent of annotating edge of the graph, while annotating output tensor is equivalent of annotating node. Thus annotation API requires quantizer to annotate nodes (output tensor) or edges (input tensors) of the graph.

For example...
```

Author: Thanks for the suggestion. It makes the introduction of ``Quantizer`` and ``QuantizationSpec`` much clearer.

1. Annotate common operator patterns
--------------------------------------------------------

In order to use the quantized pattern/operators, e.g. ``quantized add``,
backend developers will have intent to quantize (as expressed by

> `QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__

Reviewer 2: We need to explain what is quantization spec first

Author: Thanks. I think it's same as above, so add the explanation of ``QuantizationSpec`` there.

) inputs, output of the pattern. Following is an example flow (take ``add`` operator as example)
of how this intent is conveyed in the quantization workflow with annotation API.

-  Step 1: Identify the original floating point pattern in the FX graph. There are

>    several ways to identify this pattern: User may use a pattern matcher (e.g. SubgraphMatcher)

Reviewer 2: `User` => `Quantizer`

Author: Thanks and changed.

   we must annotate ``add`` as example in above section 1, in which two inputs of ``add``
   has different quantization parameters.
-  Example 2: Another example is that of sharing quantization parameters between inputs and output.
   This typically results from operators such as ``maxpool``, ``average_pool``, ``concat`` etc.

``SharedQuantizationSpec`` is designed for this use case to annotate tensors whose quantization
parameters are shared with other tensors. Input of ``SharedQuantizationSpec`` is an ``EdgeOrNode`` object which 
can be an input edge or an output value. 

-  Input edge is the connection between input node and the node consuming the input,
   so it's a ``Tuple[Node, Node]``.
-  Output value is an FX ``Node``.

Now, if we want to rewrite ``add`` annotation example with ``SharedQuantizationSpec`` to indicate
two input tensors as sharing quantization parameters. We can define its ``QuantizationAnnotation``
as this:

-  Step 1: Identify the original floating point pattern in the FX graph. We can use the same
   methods introduced in ``QuantizationSpec`` example to identify the ``add`` pattern.
-  Step 2: Annotate input_act0 of ``add`` with ``QuantizationSpec``.

> -  Step 3: Create a ``SharedQuantizationSpec`` object with input edge defined as ``(input_act0, add_node)`` which means to

Reviewer 2: THis is exactly why we need to preface this by saying that annotation API requires quantizer to annotate nodes or edges. Nodes are annotated via `output_qspec` and edges can be annotated via `input_qspec_map`. Later we will show in edge can also be explicitly annotated when quantizer wants to express intention of sharing quantization parameters or deriving them.

and run a `toy example <https://gist.github.com/leslie-fang-intel/b78ed682aa9b54d2608285c5a4897cfc>`__
with ``Torchvision Resnet18``. To better understand the final example, here are the classes and utility
functions that are used in the example:

-  `QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
   consists of ``QuantizationSpec`` for activation, weight, and bias separately.
-  When annotating the model,
   `get_input_act_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L10>`__,
   `get_output_act_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L23>`__,
   `get_weight_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L36>`__, and
   `get_bias_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L53>`__

>    can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific node.

Reviewer: ```suggestion
   can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific pattern.
```

Author: Thanks and changed.

   developer only wants to quantize inputs and outputs when the ``linear`` has a third input, it requires co-work from quantization
   team and backend developer.
-  Current API uses ``QConfigMapping`` and ``BackendConfig`` as separate object. ``QConfigMapping`` describes user's
   intention of how they want their model to be quantized. ``BackendConfig`` describes what kind of quantization a backend support.
   Currently, ``BackendConfig`` is backend specific, but ``QConfigMapping`` is not. And user can provide a ``QConfigMapping``
   that is incompatible with a specific ``BackendConfig``. This is not a great UX. Ideally, we can structure this better
   by making both configuration (``QConfigMapping``) and quantization capability (``BackendConfig``) backend
   specific. So there will be less confusion about incompatibilities.
-  Currently, in ``QConfig`` we are exposing observer/fake_quant classes as an object for user to configure quantization.
   This increases the things that user needs to care about, e.g. not only the ``dtype`` but also how the observation should

>    happen. These could potentially be hidden from user to make user interface simpler.

Reviewer 4: A bit confused by this, I think the new API also has the observer setting in the `QuantizationSpec`. So it is also part of the new API, right?

Author: Thanks for the comment. In creation of `QuantizationSpec`, we do still need to specify the class of `observer`. But I think it's simpler for users to use comparing with  `FX Quantization Mode`.

- In `FX Quantization Mode`, `QConfig` is created with observer. User needs to learn the constructor and specify the quantization parameters for different observers in order to create a `QConfig` for his use case.
- Now the general quantization parameters such as ``dtype``, ``qscheme`` for different observers are unified in the ``QuantizationSpec`` and only the observer class are needed.

@jerryzh168 I think you may help to comment more about this bullet.

Reviewer: so in the new quantizer API, the idea is that configurability is controlled by each backend, the quantizer/annotation API is not facing common modeling user directly, it is an API for backend developers or advanced modeling users only. for common users, they will be interacting with each backend specific quantizer only, e.g. I can have a quantizer that only exposes "quantize/not quantize" option to users:

```
backend_quantizer = BackendQuantizer()
backend_quantizer.enable_quantization()
model = prepare_pt2e(model, backend_quantizer)
...
```

see the graph in the end of motivation section: https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c for details

    quantizer.set_global(operator_config)
    prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)

    # Step 3: Quantize the model
    convered_graph_module = convert_pt2e(prepared_graph_module)

    # Step 4: Lower Reference Quantized Model into the backend

Quantizer uses annotation API to convey quantization intent for different operators/patterns.
Annotation API uses
`QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__
to convey intent of how a tensor will be quantized,
e.g. dtype, bitwidth, min, max values, symmetric vs. asymmetric etc.
Furthermore, annotation API also allows quantizer to specify how a
tensor value should be observed, e.g. ``MinMaxObserver``, or ``HistogramObserver``
, or some customized observer.

``QuantizationSpec`` is used to annotate nodes' input tensors or output tensor. Annotating
input tensors is equivalent of annotating edge of the graph, while annotating output tensor is
equivalent of annotating node. Thus annotation API requires quantizer to annotate

> edges (input tensors) or nodes (output tensor) of the graph.

Reviewer 4: I think it is also worth noting here about `QuantizationAnnotation` and its relationship with `QuantizationSpec`. If we view the model as a graph,  `QuantizationAnnotation` is for node (op) and `QuantizationSpec` is for edge (tensor). `QuantizationAnnotation` is composed of the `QuantizationSpec` for all the input and output tensor of an op.

perform quantization transformations on top of the captured model. In a
similar way, for Quantization 2.0 flow, we will now use the PT2 Export
workflow to capture the model into a graph, and perform quantization
transformations on top of the ATen dialect graph. This approach is expected to
have significantly higher model coverage, better programmability, and
a simplified UX.

Prerequisites:
-----------------------

-  `Understanding of torchdynamo concepts in PyTorch <https://pytorch.org/docs/stable/dynamo/index.html>`__
-  `Understanding of the quantization concepts in PyTorch <https://pytorch.org/docs/master/quantization.html#quantization-api-summary>`__
-  `Understanding of FX Graph Mode post training static quantization <https://pytorch.org/tutorials/prototype/fx_graph_mode_ptq_static.html>`__
-  `Understanding of BackendConfig in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html?highlight=backend>`__
-  `Understanding of QConfig and QConfigMapping in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html#set-up-qconfigmapping-that-satisfies-the-backend-constraints>`__

> 

Reviewer 3: nit: you can remove "Understanding of" from all these bullet points and keep them a bit more concise.

Author: Thanks for comment and removed.

> Previously in ``FX Graph Mode Quantization`` we were using ``QConfigMapping`` for users to specify how the model to be quantized

Reviewer 3: Is this a separate section? Currently it seems to fall under Prerequisites

Author: Thanks, I will add it ``Preface`` section.

   specific. So there will be less confusion about incompatibilities.
-  In ``QConfig``, we are exposing observer/fake_quant classes as an object for user to configure quantization.
   This increases the things that user needs to care about, e.g. not only the ``dtype`` but also how the
   observation should happen. These could potentially be hidden from user to make user interface simpler.

To address these scalability issues, 
`Quantizer <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L160>`__
is introduced for quantization in PyTorch 2.0 export. ``Quantizer`` is a class that users can use to
programmatically set the quantization specifications for input and output of each node in the model graph. It adds flexibility
to the quantization API and allows modeling users and backend developers to configure quantization programmatically.
This will allow users to express how they want an operator pattern to be observed in a more explicit
way by annotating the appropriate nodes. A backend specific quantizer inherited from base quantizer,
some methods that need to be implemented:

> -  `annotate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L269>`__

Reviewer 3: is this the only method that needs to be implemented (above description says some methods)? Why create a separate bullet point for it in that case?

Author: Actually there are some other methods needs to be implemented. However, the detail design for the other methods are not fully set per the discussion with @jerryzh168. So we only mention `annotate method ` here. Sure, I will merge this bullet into previous paragraph for now.

e.g. dtype, bitwidth, min, max values, symmetric vs. asymmetric etc.
Furthermore, ``QuantizationSpec`` also allows quantizer to specify how a
tensor value should be observed, e.g. ``MinMaxObserver``, or ``HistogramObserver``
, or some customized observer.

``QuantizationAnnotation`` composed of ``QuantizationSpec`` objects is used to annotate input tensors
and output tensor of a pattern. Annotating input tensors is equivalent of annotating input edges,
while annotating output tensor is equivalent of annotating node. ``QuantizationAnnotation`` is a ``dataclass``
with several fields:

-  ``input_qspec_map`` field is of class ``Dict`` to map each input tensor (as input edge) to a ``QuantizationSpec``.
-  ``output_qspec`` field expresses the ``QuantizationSpec`` used to annotate the output tensor;
-  ``_annotated`` field indicates if this node has already been annotated by quantizer.

To conclude, annotation API requires quantizer to annotate edges (input tensors) or
nodes (output tensor) of the graph. Now, we will have a step-by-step tutorial for
how to use the annotation API with different types of ``QuantizationSpec``.

> 1. Annotate common operator patterns

Reviewer 3: these seem to be individual headings. I'd recommend having a heading on Annotation API Usage and then include these within that for better formatting.

Author: Thanks and changed.

    operator_config = qq.get_symmetric_quantization_config(is_per_channel=True)
    quantizer.set_global(operator_config)
    prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)

    # Step 3: Quantize the model
    convered_graph_module = convert_pt2e(prepared_graph_module)

    # Step 4: Lower Reference Quantized Model into the backend

Inside the Quantizer, we will use the ``QuantizationAnnotation API``
to convey user's intent for what quantization spec to use and how to
observe certain tensor values in the prepare step. Now, we will have a step-by-step
tutorial for how to use the ``QuantizationAnnotation API`` with different types of
``QuantizationSpec``.

1. Annotate common operator patterns
--------------------------------------------------------

`QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__
is used to annotate common operators like ``conv2d``. It allows user to specify how to quantize
input tensors and output tensor of this operator which includes parameters of ``observer type``, ``dtype``,
``quant_min``, and ``quant_max`` etc.

> ::

Reviewer: I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:
1. identify conv2d pattern
2. derive the quantization spec for input and output Tensors
3. annotate the nodes in the pattern properly

you can also talk about how to annotate conv - relu pattern in this section

Reviewer: maybe start with add / add - relu as an example here, since conv requires DerivedQuantizationSpec for bias

Reviewer: you can follow the steps in https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#heading=h.it9h4gjr7m9g for each section that introduces a new quantization spec I think

Author: > I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:

Hi @jerryzh168, Thanks for the suggestions. I have updated the document by elaboration more about use case and steps to use each ``QuantizationSpec``.  Please help to take a look again.
One question is about ``DerivedQuantizationSpec``, I think this [document ](https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#) didn't introduce motivation clearly. Since we generally use `PlaceholderObserver` with `QuantizationSpec` for bias node, why we have to use ``DerivedQuantizationSpec``?

Reviewer: I see, this is for the Tensors whose quantization parameters are derived from other Tensors, for example bias, the `PlaceholderObserver` that we used previously did not express this relation correctly

Author: Hi @jerryzh168. Thanks for the comment. Will the [`bias_quantization_spec`](https://github.com/pytorch/pytorch/blob/a02a58d862b00a7aa410149e8cf276b8d0ca3462/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L183-L185) in `QNNPackQuantizer` updated to use `DerivedQuantizationSpec`? If so, does it mean there will be the `fp32_bias -> quant_to_int32 -> dequant -> conv` pattern instead of `fp32_bias -> conv` after `convert_pt2e`?

Reviewer: yeah I think it will be updated eventually, but I don't know when we plan to do that yet, @kimishpatel would have more context on the timeline

            )

            bias_observer_or_fake_quant_ctr: _ObserverOrFakeQuantizeConstructor = PlaceholderObserver
            bias_quantization_spec = QuantizationSpec(
                dtype=torch.float,
                observer_or_fake_quant_ctr=bias_observer_or_fake_quant_ctr
            )

            input_qspec_map = {}
            input_act = conv_node.args[0]
            assert isinstance(input_act, Node)
            input_qspec_map[input_act] = act_quantization_spec

            weight = conv_node.args[1]
            assert isinstance(weight, Node)
            input_qspec_map[weight] = weight_quantization_spec

            bias = conv_node.args[2]
            if isinstance(bias, Node):
                input_qspec_map[bias] = bias_quantization_spec

            conv_node.meta["quantization_annotation"] = QuantizationAnnotation(
                input_qspec_map=input_qspec_map,
                output_qspec=get_act_qspec(quantization_config),
                _annotated=True,
            )

2. Annotate sharing qparams operators
--------------------------------------------------------

> `SharedQuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L90>`__

Reviewer: similarly I think it's better to start with the use case of what quantized add/cat expects for its input and what the output would look like for quantized add, then derive the quantization spec for input/outputs and describe how to annotate the nodes in the add/cat pattern

Author: Sure, changed accordingly.

Today we have `FX Graph Mode
Quantization <https://pytorch.org/docs/stable/quantization.html#prototype-fx-graph-mode-quantization>`__
which uses ``symbolic_trace`` to capture the model into a graph, and then
perform quantization transformations on top of the captured model. In a
similar way, for Quantization 2.0 flow, we will now use the PT2 Export
workflow to capture the model into a graph, and perform quantization
transformations on top of the ATen dialect graph. This approach is expected to
have significantly higher model coverage, better programmability, and
a simplified UX.

Prerequisites:
-----------------------

-  `Understanding of the quantization concepts in PyTorch <https://pytorch.org/docs/master/quantization.html#quantization-api-summary>`__
-  `Understanding of FX graph mode post training static quantization <https://pytorch.org/tutorials/prototype/fx_graph_mode_ptq_static.html>`__
-  `Understanding of torchdynamo concepts in PyTorch <https://pytorch.org/docs/stable/dynamo/index.html>`__

Imagine a backend developer who wishes to integrate a third-party backend
with PyTorch's quantization 2.0 flow. To accomplish this, they would only need

> to define the backend specific quantizer. The high level architecture of

Reviewer: I think we would need to mention what quantizer is here

Author: Any suggestions about how to introduce quantizer API of `get_supported_operators`, as we discussed here https://github.com/pytorch/tutorials/pull/2354#discussion_r1213834401. It's not finalized yet, but it's a must have when define a `Quantizer`.

Reviewer: I think you can skip explaining that here, please focus on the role of Quantizer and the annotate function, you can also try to have a quick summarize of https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c here when talking about the motivation of the quantizer API

Author: Thanks. Add the introduction of ``Quantizer`` such as why we introduce this class and its APIs.

`QNNPackQuantizer <https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py>`__.
Taking QNNPackQuantizer as an example, the overall Quantization 2.0 flow could be:

::

    import torch
    import torch._dynamo as torchdynamo
    from torch.ao.quantization._quantize_pt2e import convert_pt2e, prepare_pt2e
    import torch.ao.quantization._pt2e.quantizer.qnnpack_quantizer as qq

    class M(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.linear = torch.nn.Linear(5, 10)

        def forward(self, x):
            return self.linear(x)

    example_inputs = (torch.randn(1, 5),)
    model = M().eval()

    # Step 1: Trace the model into an FX graph of flattened ATen operators
    exported_graph_module, guards = torchdynamo.export(
        model,
        *copy.deepcopy(example_inputs),
        aten_graph=True,
    )

    # Step 2: Insert observers or fake quantize modules
    quantizer = qq.QNNPackQuantizer()
    operator_config = qq.get_symmetric_quantization_config(is_per_channel=True)
    quantizer.set_global(operator_config)

>     prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)

Reviewer: can you add a comment here saying `prepare_pt2e_quantizer` will be updated to `prepare_pt2e` soon?

Author: Thanks and add the note.

    # Step 3: Quantize the model
    convered_graph_module = convert_pt2e(prepared_graph_module)

    # Step 4: Lower Reference Quantized Model into the backend

Inside the Quantizer, we will use the ``QuantizationAnnotation API``
to convey user's intent for what quantization spec to use and how to
observe certain tensor values in the prepare step. Now, we will have a step-by-step
tutorial for how to use the ``QuantizationAnnotation API`` with different types of
``QuantizationSpec``.

1. Annotate common operator patterns
--------------------------------------------------------

In order to use the quantized operators, e.g. ``quantized add``,
backend developers will have intent to quantize (as expressed by
`QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__
) input, output of the operator. Following is an example flow (with ``add``)
of how this intent is conveyed in the quantization workflow with node annotation API.

> -  Step1: Identify the original floating point ``add`` node in the FX graph. There are

Reviewer: nit: space between Step and 1. same for other things

Reviewer: also we should be identifying a pattern (e.g. linear pattern) instead of just a node

Author: Thanks. Added.

   several ways to identify this node: 1. User may use a pattern matcher (e.g. SubgraphMatcher)
   to match the operator pattern. 2. User may go through the nodes from start to the end and compare
   the node's target type.
-  Step2: Define the ``QuantizationSpec`` for two inputs and one output of the ``add`` node to specify

>    how to quantize input tensors and output tensor which includes parameters of ``observer type``,

Reviewer: I think it would be helpful to explain what ``observer type`` is, it's actually a constructor, not really a type, it can serve two purposes,
(1) allow users to pass around additional parameters on top of the fields specified in QuantizationSpec, e.g. MinMaxObserver(eps=...)
(2) allow users to have custom observer or fake quantizers

Author: Thanks for the suggestion. After refactor, this part has been removed.

   ``dtype``, ``quant_min``, and ``quant_max`` etc.

> -  Step3: Annotate the inputs and output of the ``add`` node. User will create the ``QuantizationAnnotation``

Reviewer: I think it would be helpful to explain the structure of QuantizationAnnotation here as well, i.e. what input_qspec_map and output_qspec means, and potentially give an example here about two nodes connected together, e.g.
conv1 -> conv2, how would we annotate both to int8 and what would quant flow do in this case for the output of conv1/input of conv2

Author: Thanks. ``QuantizationAnnotation`` is added.

   object and add it into ``add`` node's ``meta`` property.

::

>     def _annotate_add(

Reviewer: I feel we do not need to paste this whole code, the tutorial should focus on explaining every line of code instead, we can extract the relevant part and put that in each section and have good explanations for them so that users can take this and implement their own annotation

Reviewer: maybe some of the explanations can be in the docstrings, not necessarily here, we don't need to go into a lot of details but mostly focus on explaining how these things are connected here

- `QuantizationSpec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L28-L66>`__
  defines the ``data type``, ``qscheme``, and other quantization parameters used to quantize a tensor.
- `QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
  consists of ``QuantizationSpec`` for activation, weight, and bias separately.
- When annotating the model, methods of
  `get_act_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#L9>`__,
  `get_weight_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#L26>`__, and
  `get_bias_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#LL42C5-L42C19>`__
  can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific node.

> .. code:: ipython3

Reviewer: please remove this code as well

Author: Thanks this code are removed and add a gist link here.

have significantly higher model coverage, better programmability, and
a simplified UX.

Prerequisites:
-----------------------

-  `Understanding of the quantization concepts in PyTorch <https://pytorch.org/docs/master/quantization.html#quantization-api-summary>`__
-  `Understanding of FX graph mode post training static quantization <https://pytorch.org/tutorials/prototype/fx_graph_mode_ptq_static.html>`__
-  `Understanding of torchdynamo concepts in PyTorch <https://pytorch.org/docs/stable/dynamo/index.html>`__

Imagine a backend developer who wishes to integrate a third-party backend
with PyTorch's quantization 2.0 flow. To accomplish this, they would only need
to define the backend specific quantizer. The high level architecture of
quantization 2.0 with quantizer could look like this:

::

    float_model(Python)                               Input
        \                                              /
         \                                            /
    —-------------------------------------------------------
    |                    Dynamo Export                     |
    —-------------------------------------------------------
                                |

>                         FX Graph in CATen    QNNPackQuantizer,

Author: Thanks and Changed.

            self.operator_type_config: Dict[str, Optional[QuantizationConfig]] = {}

        def set_global(self, quantization_config: QuantizationConfig):
            """set global QuantizationConfig used for the backend.
            QuantizationConfig is defined in torch/ao/quantization/_pt2e/quantizer/quantizer.py.
            """
            self.global_config = quantization_config
            return self

        def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
            """annotate nodes in the graph with observer or fake quant constructors
            to convey the desired way of quantization.
            """
            global_config = self.global_config
            self.annotate_symmetric_config(model, global_config)

            return model

        def annotate_symmetric_config(
            self, model: torch.fx.GraphModule, config: QuantizationConfig
        ) -> torch.fx.GraphModule:
            for node in reversed(model.graph.nodes):
                # The annotation methods for each op will defined later
                pass
            return model

        def validate(self, model: torch.fx.GraphModule) -> None:

>             """validate the annotated graph is supported by the backend"""

Author: Thanks and changed.

            pass

        @classmethod
        def get_supported_operators(cls) -> List[OperatorConfig]:

>             """return the operator list which is supported by the backend"""

Author: Add my understanding about these questions:

- The semantic of `get_supported_operators`.
- The semantic of `OperatorConfig`.

Hi @jerryzh168, I think we need your help here for further elaboration about `get_supported_operators` since it's a  abstractmethod in base `Quantizer` which must be implemented by BackendQuantizer.

>             """return the operator list which is supported by the backend"""

Reviewer: I don't think these need to be in this doc, the main goal for this doc is to introduce different types of quantization specs and how to annotate patterns

>             """return the operator list which is supported by the backend"""

Author: Got it, but we still need to define this method `get_supported_operators` for final toy example since it's an abstractmethod in base Quantizer which must be implemented by BackendQuantizer. So, did you suggest we skip this section and directly keep it in the final toy example?
